src/java/org/apache/nutch/util/SitemapProcessor.java

```diff
@@ -0,0 +1,436 @@
+          if(!strict || sitemapUrl.isValid()) {
+            String key = filterNormalize(sitemapUrl.getUrl().toString());
+            if (key != null) {
+              CrawlDatum sitemapUrlDatum = new CrawlDatum();
+              sitemapUrlDatum.setStatus(CrawlDatum.STATUS_SITEMAP);
+              sitemapUrlDatum.setScore((float) sitemapUrl.getPriority());
+
+              if(sitemapUrl.getChangeFrequency() != null) {
+                int fetchInterval = -1;
+                switch(sitemapUrl.getChangeFrequency()) {
+                  case ALWAYS:  fetchInterval = 1;        break;
+                  case HOURLY:  fetchInterval = 3600;     break; // 60*60
+                  case DAILY:   fetchInterval = 86400;    break; // 60*60*24
+                  case WEEKLY:  fetchInterval = 604800;   break; // 60*60*24*7
+                  case MONTHLY: fetchInterval = 2592000;  break; // 60*60*24*30
+                  case YEARLY:  fetchInterval = 31536000; break; // 60*60*24*365
+                  case NEVER:   fetchInterval = Integer.MAX_VALUE; break; // Loose "NEVER" contract
+                }
+                sitemapUrlDatum.setFetchInterval(fetchInterval);
+              }
+
+              if(sitemapUrl.getLastModified() != null)
+                sitemapUrlDatum.setModifiedTime(sitemapUrl.getLastModified().getTime());
+
+              context.write(new Text(key), sitemapUrlDatum);
+            }
+          }
+        }
+      }
+      else if (asm instanceof SiteMapIndex) {
+        SiteMapIndex index = (SiteMapIndex) asm;
+        Collection<AbstractSiteMap> sitemapUrls = index.getSitemaps();
+
+        for(AbstractSiteMap sitemap: sitemapUrls) {
+          if(sitemap.isIndex()) {
+            generateSitemapUrlDatum(protocol, sitemap.getUrl().toString(), context);
+          }
+        }
+      }
+    }
+  }
+
+  private static class SitemapReducer extends Reducer<Text, CrawlDatum, Text, CrawlDatum> {
+    CrawlDatum sitemapDatum  = null;
+    CrawlDatum originalDatum = null;
+
+    public void reduce(Text key, Iterable<CrawlDatum> values, Context context)
+        throws IOException, InterruptedException {
+      sitemapDatum  = null;
+      originalDatum = null;
+
+      for (CrawlDatum curr: values) {
+        if(curr.getStatus() == CrawlDatum.STATUS_SITEMAP && sitemapDatum == null) {
+          sitemapDatum = new CrawlDatum();
+          sitemapDatum.set(curr);
+        }
+        else {
+          originalDatum = new CrawlDatum();
+          originalDatum.set(curr);
+        }
+      }
+
+      if(originalDatum != null) {
+        // The url was already present in crawldb. If we got the same url from sitemap too, save
+        // the information from sitemap to the original datum. Emit the original crawl datum
+        if(sitemapDatum != null) {
+          context.getCounter("Sitemap", "existing_sitemap_entries").increment(1);
+          originalDatum.setScore(sitemapDatum.getScore());
```

Comment:
- the [sitemap spec](https://www.sitemaps.org/protocol.html#xmlTagDefinitions) defines "the priority of this URL relative to other URLs on your site." That's different from a global score as calculated by OPIC, page rank, etc.
- overwriting fetchInterval will make [calculateLastFetchTime()](https://github.com/apache/nutch/blob/master/src/java/org/apache/nutch/crawl/AbstractFetchSchedule.java#L158) return the wrong time when a page has been fetched
- overwriting the modified time breaks any if-modified-since handling.

See the discussion in [NUTCH-1465](https://issues.apache.org/jira/browse/NUTCH-1465) on Jan 30-31, 2014. I was also wrong how to map these "concepts" from sitemaps to Nutch-internal ones.